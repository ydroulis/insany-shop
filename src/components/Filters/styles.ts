"use client"

import styled from 'styled-components'

export const Container = styled.section`
    margin: 0 auto; 
    max-width: 1120px; 
    display: flex; 
    alignItems: center; 
    justify-content: space-between; 
    margin-top: 2rem;

    ul {
        display: flex;
        gap: 4px;
        list-style: none;
        font-size: 0.875rem;
        color: #737380;
    }

    a, .not-active {
        text-decoration: none;
        color: #737380;

        &:hover {
            text-decoration: underline;
        }
    }

    .active {
        text-decoration: none;
        color: #000000;

        &:hover {
            text-decoration: underline;
        }
    }
`