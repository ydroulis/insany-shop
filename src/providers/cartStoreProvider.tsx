'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CartStore, createCartStore } from '../stores/cartStore'

export type CartStoreApi = ReturnType<typeof createCartStore>

export const CartStoreContext = createContext<CartStoreApi | undefined>(
    undefined,
)

export interface CartStoreProviderProps {
    children: ReactNode
}

export const CartStoreProvider = ({
    children,
}: CartStoreProviderProps) => {
    const storeRef = useRef<CartStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCartStore()
    }

    return (
        <CartStoreContext.Provider value={storeRef.current}>
            {children}
        </CartStoreContext.Provider>
    )
}

export const useCartStore = <T,>(
    selector: (store: CartStore) => T,
): T => {
    const cartStoreContext = useContext(CartStoreContext)

    if (!cartStoreContext) {
        throw new Error(`useStore must be used within CartStoreProvider`)
    }

    return useStore(cartStoreContext, selector)
}
