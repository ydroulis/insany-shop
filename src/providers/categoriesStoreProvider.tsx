'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CategoriesStore, createCategoriesStore } from '../stores/categoriesStore'

export type CategoriesStoreApi = ReturnType<typeof createCategoriesStore>

export const CategoriesStoreContext = createContext<CategoriesStoreApi | undefined>(
    undefined,
)

export interface CategoriesStoreProviderProps {
    children: ReactNode
}

export const CategoriesStoreProvider = ({
    children,
}: CategoriesStoreProviderProps) => {
    const storeRef = useRef<CategoriesStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCategoriesStore()
    }

    return (
        <CategoriesStoreContext.Provider value={storeRef.current}>
            {children}
        </CategoriesStoreContext.Provider>
    )
}

export const useCategoriesStore = <T,>(
    selector: (store: CategoriesStore) => T,
): T => {
    const categoriesStoreContext = useContext(CategoriesStoreContext)

    if (!categoriesStoreContext) {
        throw new Error(`useStore must be used within CategoriesStoreProvider`)
    }

    return useStore(categoriesStoreContext, selector)
}
