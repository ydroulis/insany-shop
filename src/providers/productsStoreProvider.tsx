'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type ProductsStore, createProductsStore } from '../stores/productsStore'

export type ProductsStoreApi = ReturnType<typeof createProductsStore>

export const ProductsStoreContext = createContext<ProductsStoreApi | undefined>(
    undefined,
)

export interface ProductsStoreProviderProps {
    children: ReactNode
}

export const ProductsStoreProvider = ({
    children,
}: ProductsStoreProviderProps) => {
    const storeRef = useRef<ProductsStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createProductsStore()
    }

    return (
        <ProductsStoreContext.Provider value={storeRef.current}>
            {children}
        </ProductsStoreContext.Provider>
    )
}

export const useProductsStore = <T,>(
    selector: (store: ProductsStore) => T,
): T => {
    const productsStoreContext = useContext(ProductsStoreContext)

    if (!productsStoreContext) {
        throw new Error(`useStore must be used within ProductsStoreProvider`)
    }

    return useStore(productsStoreContext, selector)
}
