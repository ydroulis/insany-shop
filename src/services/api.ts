import axios from "axios";

export const api = axios.create({
    baseURL: "https://api.insany.co/api",
    headers: {
        "Content-Type": "application/json",
    },
});