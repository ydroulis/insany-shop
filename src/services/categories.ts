import { Categories } from "@/types/Categories";
import { api } from "./api";


export const getCategories = async (): Promise<{ categories: Categories }> => {
    const response = await api.get<{ categories: Categories }>("/categories");
    return response.data;
};
