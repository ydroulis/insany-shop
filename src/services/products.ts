import { Product, ProductsResponse } from "@/types/Products";
import { api } from "./api";

export const getProducts = async (
    page?: number,
    limit?: number,
    category?: string
): Promise<ProductsResponse> => {
    const params: Record<string, string | number> = {};
    if (page) params.page = page;
    if (limit) params.limit = limit;
    if (category) params.category = category;

    const response = await api.get<ProductsResponse>("/products", { params });
    return response.data;
};

export const getProductById = async (id: number): Promise<Product> => {
    const response = await api.get<{ product: Product }>(`/products/${id}`);
    return response.data.product;
};
