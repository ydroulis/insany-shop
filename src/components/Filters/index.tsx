"use client"
import React, { useEffect, useState } from 'react';

import * as S from './styles';
import Selector from '../Selector';
import Link from 'next/link';
import { usePathname } from 'next/navigation';
import { useCategoriesStore } from '../../providers/categoriesStoreProvider';
import { useProductsStore } from '../../providers/productsStoreProvider';
import { getProducts } from '../../services/products';

interface FiltersProps {
    pageId?: string
}

const optionsCategory = [
    { value: 'eletronicos', label: 'Eletrônicos' },
    { value: 'roupas', label: 'Roupas e Calçados' },
    { value: 'casa', label: 'Casa e Decoração' },
    { value: 'livros', label: 'Livros' },
    { value: 'esportes', label: 'Esporte e Lazer' },
]

const optionsSort = [
    { value: 'newers', label: 'Novidades' },
    { value: 'older', label: 'Mais Antigos' },
    { value: 'higher', label: 'Preço: Maior - menor' },
    { value: 'lower', label: 'Preço: Menor - maior' },
]

const Filters: React.FC<FiltersProps> = ({ pageId }) => {
    const [valueCategory, setValueCategory] = useState('Selecione a categoria');
    const [valueSort, setValueSort] = useState('Organizar por');

    const { categories } = useCategoriesStore((state) => state);
    const { products, setProducts, setPagination } = useProductsStore((state) => state);




    useEffect(() => {
        const categoryValueSelected = optionsCategory.find(option => option.label === valueCategory)?.value

        if (valueCategory !== 'Selecione a categoria') {
            const fetchCategoryProducts = async () => {
                const res = await getProducts(1, 100, categoryValueSelected);
                setPagination(res.pagination);
                setProducts(res.products);
            }
            fetchCategoryProducts()
        }

    }, [setPagination, setProducts, valueCategory])

    useEffect(() => {
        switch (valueSort) {
            case 'Novidades':
                setProducts(products.sort((a, b) => b.id - a.id));
                break;
            case 'Mais Antigos':
                setProducts(products.sort((a, b) => a.id - b.id));
                break;
            case 'Preço: Maior - menor':
                setProducts(products.sort((a, b) => b.price - a.price));
                break;
            case 'Preço: Menor - maior':
                setProducts(products.sort((a, b) => a.price - b.price));
                break;
            default:
                break;
        }
    }, [products, setProducts, valueSort])

    const pathname = usePathname()
    const isActive = pathname === `/category/${pageId}`
    const linkClassname = isActive ? 'active' : 'not-active'

    const category = categories.find(category => category.id === pageId)

    return (
        <>
            {products.length > 0 && (
                <S.Container data-testid="filters-section">
                    {pageId ? (
                        <nav aria-label="breadcrumb">
                            <ul>
                                <li>
                                    <Link aria-label='Ir para a página inicial' href="/" >Produtos</Link>
                                </li>
                                <span aria-hidden="true"> / </span>
                                <li>
                                    <Link
                                        aria-label={`Ir para a categoria ${category?.name}`}
                                        className={linkClassname}
                                        href={`/category/${pageId}`}
                                        aria-current={isActive ? "page" : undefined}
                                    >
                                        {category?.name}
                                    </Link>
                                </li>
                            </ul>
                        </nav>
                    ) : (

                        <Selector
                            type="category"
                            options={optionsCategory}
                            setValue={setValueCategory}
                            value={valueCategory}
                            label="Selecione a categoria" id="category"
                        />
                    )}
                    <Selector
                        type="sort"
                        options={optionsSort}
                        setValue={setValueSort}
                        value={valueSort}
                        label="Selecione a organização dos produtos" id="sort"
                    />
                </S.Container>
            )}
        </>
    );
}

export default Filters;