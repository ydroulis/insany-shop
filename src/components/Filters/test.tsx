import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Filters from '../Filters';
import { useCategoriesStore } from '../../providers/categoriesStoreProvider';
import { useProductsStore } from '../../providers/productsStoreProvider';
import { usePathname } from 'next/navigation';
import '@testing-library/jest-dom';
import { getProducts } from '../../services/products';

jest.mock('../../providers/categoriesStoreProvider');
jest.mock('../../providers/productsStoreProvider');
jest.mock('next/navigation', () => ({
    usePathname: jest.fn(),
}));
jest.mock('../../services/products');

describe('Filters Component', () => {
    const mockCategories = [
        { id: 'eletronicos', name: 'Eletrônicos' },
        { id: 'clothing', name: 'Roupas e Calçados' },
    ];

    const mockProducts = [
        { id: 1, name: 'Produto 1', price: 100 },
        { id: 2, name: 'Produto 2', price: 50 },
    ];

    const mockSetProducts = jest.fn();

    beforeEach(() => {
        (useCategoriesStore as jest.Mock).mockReturnValue({ categories: mockCategories });
        (useProductsStore as jest.Mock).mockReturnValue({ products: mockProducts, setProducts: mockSetProducts });
        (usePathname as jest.Mock).mockReturnValue('/');
        (getProducts as jest.Mock).mockResolvedValue({ products: mockProducts });
        jest.clearAllMocks();
    });

    it('renders selectors when no pageId is passed', () => {
        render(<Filters />);

        const filtersSection = screen.getByTestId('filters-section');
        expect(filtersSection).toBeInTheDocument();

        expect(screen.getByLabelText('Selecione a categoria')).toBeInTheDocument();
        expect(screen.getByLabelText('Selecione a organização dos produtos')).toBeInTheDocument();
    });

    it('renders breadcrumbs when pageId is passed', () => {
        (usePathname as jest.Mock).mockReturnValue('/category/eletronicos');

        render(<Filters pageId="eletronicos" />);

        const nav = screen.getByRole('navigation', { name: /breadcrumb/i });
        expect(nav).toBeInTheDocument();

        expect(screen.getByText('Produtos')).toBeInTheDocument();
        const categoryLink = screen.getByText('Eletrônicos');
        expect(categoryLink).toBeInTheDocument();

        expect(categoryLink).toHaveClass('active');
        expect(categoryLink).toHaveAttribute('aria-current', 'page');
    });

    it('applies not-active class when breadcrumb link is not active', () => {
        (usePathname as jest.Mock).mockReturnValue('/');

        render(<Filters pageId="eletronicos" />);
        const categoryLink = screen.getByText('Eletrônicos');
        expect(categoryLink).toHaveClass('not-active');
        expect(categoryLink).not.toHaveAttribute('aria-current');
    });

    it('handles category not found in breadcrumbs', () => {
        render(<Filters pageId="unknown" />);
        const categoryLink = screen.queryByText('unknown');
        expect(categoryLink).not.toBeInTheDocument();
    });

    it('updates category selector value and fetches products', async () => {
        render(<Filters />);

        const categoryButton = screen.getByLabelText('Selecione a categoria');
        fireEvent.click(categoryButton);

        const option = screen.getByRole('option', { name: 'Eletrônicos' });
        fireEvent.click(option);

        await waitFor(() => {
            expect(mockSetProducts).toHaveBeenCalledWith(mockProducts);
        });

        expect(categoryButton).toHaveTextContent('Eletrônicos');
    });

    it('updates sort selector value and sorts products', async () => {
        render(<Filters />);

        const sortButton = screen.getByLabelText('Selecione a organização dos produtos');
        fireEvent.click(sortButton);

        const option = screen.getByRole('option', { name: 'Preço: Maior - menor' });
        fireEvent.click(option);

        await waitFor(() => {
            expect(mockSetProducts).toHaveBeenCalledWith([...mockProducts].sort((a, b) => b.price - a.price));
        });

        expect(sortButton).toHaveTextContent('Preço: Maior - menor');
    });

    it('does not render filters section if there are no products', () => {
        (useProductsStore as jest.Mock).mockReturnValue({ products: [], setProducts: mockSetProducts });

        render(<Filters />);

        expect(screen.queryByTestId('filters-section')).not.toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { container } = render(<Filters />);
        expect(container).toMatchSnapshot();
    });
});
