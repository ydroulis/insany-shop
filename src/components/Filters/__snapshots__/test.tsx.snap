// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Filters Component matches snapshot 1`] = `
<div>
  <section
    class="sc-bRKDuR hyYVIi"
    data-testid="filters-section"
  >
    <div
      class="sc-hvigdm cfRGDx"
    >
      <button
        aria-controls="selector-menu"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Selecione a categoria"
        class="sc-fhHczv YbXzB"
        id="category-selector"
      >
        <span>
          Selecione a categoria
        </span>
        <svg
          aria-hidden="true"
          class="icon"
          color="#737380"
          fill="currentColor"
          height="16"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(115, 115, 128);"
          viewBox="0 0 448 512"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </button>
      <div
        aria-labelledby="custom-selector"
        class="sc-ggWZvA bSoREI"
        id="category-menu"
        role="listbox"
      >
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Eletrônicos
        </button>
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Roupas e Calçados
        </button>
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Casa e Decoração
        </button>
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Livros
        </button>
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Esporte e Lazer
        </button>
      </div>
    </div>
    <div
      class="sc-hvigdm cfRGDx"
    >
      <button
        aria-controls="selector-menu"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Selecione a organização dos produtos"
        class="sc-fhHczv ikLxHS"
        id="sort-selector"
      >
        <span>
          Organizar por
        </span>
        <svg
          aria-hidden="true"
          class="icon"
          color="#737380"
          fill="currentColor"
          height="16"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(115, 115, 128);"
          viewBox="0 0 448 512"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </button>
      <div
        aria-labelledby="custom-selector"
        class="sc-ggWZvA bSoREI"
        id="sort-menu"
        role="listbox"
      >
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Novidades
        </button>
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Mais Antigos
        </button>
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Preço: Maior - menor
        </button>
        <button
          aria-selected="false"
          class="sc-dTvVRJ irVMIK"
          role="option"
          tabindex="0"
        >
          Preço: Menor - maior
        </button>
      </div>
    </div>
  </section>
</div>
`;
